### PR TITLE
DDF-2986 Resolves timing issue with MetadataConfigurationParser callbacks

### PR DESCRIPTION
**Pull request by coyotesqrl**
#### What does this PR do?
Resolves timing issue with idp server and adds callback for idp client.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@brendan-hofmann @bdeining 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[Security](https://github.com/orgs/codice/teams/security)
@stustison 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@clockard
@lessarderic

#### How should this be tested? (List steps with links to updated documentation)
Anyone who can successfully duplicate the idp login issue should attempt to login with this update. As we have not found a reproducible way to cause the timing issue, there is no definitive test.

#### Any background context you want to provide?
N/A

#### What are the relevant tickets?
[DDF-2986](https://codice.atlassian.net/browse/DDF-2986)

#### Screenshots (if appropriate)
N/A

#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java [hunk 1]
```diff
@@ -61,8 +62,9 @@
     public void setMetadata(String metadata)
             throws WSSecurityException, XMLStreamException, SAMLException, IOException {
         MetadataConfigurationParser metadataConfigurationParser = new MetadataConfigurationParser(
-                Collections.singletonList(metadata));
-        entryDescriptions = metadataConfigurationParser.getEntryDescriptions();
+                Collections.singletonList(metadata),
+                ed -> entryDescriptions.put(ed.getEntityID(), ed));
```
Comment (stustison) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java]: Do we need the same null check here that you added for the server?

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java [hunk 2]
```diff
@@ -82,8 +82,8 @@ public MetadataConfigurationParser(List<String> entityDescriptions) throws IOExc
 
     public MetadataConfigurationParser(List<String> entityDescriptions,
             Consumer<EntityDescriptor> updateCallback) throws IOException {
-        parseEntityDescriptions(entityDescriptions);
         this.updateCallback = updateCallback;
+        parseEntityDescriptions(entityDescriptions);
```
Comment (stustison) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/MetadataConfigurationParser.java]: Nice catch, I must have looked at that 50 times and didn't see it lol

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java (hunk 1 above)
Comment (coyotesqrl) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java]: Not that I can see. On the server side, we're building a new `EntityInformation` with the `EntityDescriptor` and the `build` method on that can return null. The ED shouldn't ever be null here.

**Comment by coyotesqrl** [comment]
build now

**Pull request merged**